modules/phase_field/src/ics/MultiSmoothSuperellipsoidIC.C

```diff
@@ -95,29 +95,48 @@ MultiSmoothSuperellipsoidIC::computeSuperellipsoidSemiaxes()
 {
   for (unsigned int i = 0; i < _numbub[_gk]; i++)
   {
-    //Vary bubble radius
     switch (_semiaxis_variation_type)
     {
-    case 0: //Uniform distrubtion
-      _as.push_back(_semiaxis_a[_gk]*(1.0 + (1.0 - 2.0 * _random.rand(_tid)) * _semiaxis_a_variation[_gk]));
-      _bs.push_back(_semiaxis_b[_gk]*(1.0 + (1.0 - 2.0 * _random.rand(_tid)) * _semiaxis_b_variation[_gk]));
-      _cs.push_back(_semiaxis_c[_gk]*(1.0 + (1.0 - 2.0 * _random.rand(_tid)) * _semiaxis_c_variation[_gk]));
-      break;
-    case 1: //Normal distribution
-      _as.push_back(_random.randNormal(_tid, _semiaxis_a[_gk],_semiaxis_a_variation[_gk]));
-      _bs.push_back(_random.randNormal(_tid, _semiaxis_b[_gk],_semiaxis_b_variation[_gk]));
-      _cs.push_back(_random.randNormal(_tid, _semiaxis_c[_gk],_semiaxis_c_variation[_gk]));
-      break;
-    case 2: //No variation
-      _as.push_back(_semiaxis_a[_gk]);
-      _bs.push_back(_semiaxis_b[_gk]);
-      _cs.push_back(_semiaxis_c[_gk]);
-    }
+      case 0: //Uniform distrubtion
+        if (_vary_axes_independently)
+        {
+          _as.push_back(_semiaxis_a[_gk] * (1.0 + (1.0 - 2.0 * _random.rand(_tid)) * _semiaxis_a_variation[_gk]));
+          _bs.push_back(_semiaxis_b[_gk] * (1.0 + (1.0 - 2.0 * _random.rand(_tid)) * _semiaxis_b_variation[_gk]));
+          _cs.push_back(_semiaxis_c[_gk] * (1.0 + (1.0 - 2.0 * _random.rand(_tid)) * _semiaxis_c_variation[_gk]));
+        }
+        else
+        {
+          const Real randnum = _random.rand(_tid);
+          _as.push_back(_semiaxis_a[_gk] * (1.0 + (1.0 - 2.0 * randnum) * _semiaxis_a_variation[_gk]));
+          _bs.push_back(_semiaxis_b[_gk] * (1.0 + (1.0 - 2.0 * randnum) * _semiaxis_b_variation[_gk]));
+          _cs.push_back(_semiaxis_c[_gk] * (1.0 + (1.0 - 2.0 * randnum) * _semiaxis_c_variation[_gk]));
+        }
+        break;
 
-    if (_as.back() < 0.0) _as.back() = 0.0;
-    if (_bs.back() < 0.0) _bs.back() = 0.0;
-    if (_cs.back() < 0.0) _cs.back() = 0.0;
+      case 1: //Normal distribution
+        if (_vary_axes_independently)
+        {
+          _as.push_back(_random.randNormal(_tid, _semiaxis_a[_gk],_semiaxis_a_variation[_gk]));
+          _bs.push_back(_random.randNormal(_tid, _semiaxis_b[_gk],_semiaxis_b_variation[_gk]));
+          _cs.push_back(_random.randNormal(_tid, _semiaxis_c[_gk],_semiaxis_c_variation[_gk]));
+        }
+        else
+        {
+          const Real randnum = _random.randNormal(_tid, 0, 1);
+          _as.push_back(_semiaxis_a[_gk] + (randnum * _semiaxis_a_variation[_gk]));
+          _bs.push_back(_semiaxis_b[_gk] + (randnum * _semiaxis_b_variation[_gk]));
+          _cs.push_back(_semiaxis_c[_gk] + (randnum * _semiaxis_c_variation[_gk]));
+        }
+        break;
 
+      case 2: //No variation
+        _as.push_back(_semiaxis_a[_gk]);
+        _bs.push_back(_semiaxis_b[_gk]);
+        _cs.push_back(_semiaxis_c[_gk]);
```

Comment:
I can do that, it just seemed like a headache to keep track of which index I need to write to across each set of shapes it is generating instead of just adding them at the end as I went. I am thinking I can use the global index _gk to work out exactly which index i need inside the loops of computerSuperellipsoidCenters, computeSuperellipsoidExponents, and computeSuperellipsoidSemiaxes.

Is there a reason push_back is undesirable?